reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+                            .toSet();
+                    if (rule.conclusion().generating().isPresent() && ruleSideVariables.contains(rule.conclusion().generating().get())) {
+                        // There is one generated variable per combination of ALL variables in the conclusion
+                        ruleSideVariables = rule.conclusion().pattern().variables().stream().filter(Variable::isThing).collect(Collectors.toSet());
+                    }
+                    inferredEstimate += answerCountEstimator.estimateAnswers(rule.condition().conjunction(), ruleSideVariables);
+                }
+            }
+            return inferredEstimate;
+        }
+
+        private ConjunctionAnswerCountEstimator.LocalEstimate estimatesFromHasEdges(Concludable.Has concludableHas) {
+            long estimate = countPersistedHasEdges(concludableHas.asHas().owner().inferredTypes(), concludableHas.asHas().attribute().inferredTypes()) +
+                    estimateInferredAnswerCount(concludableHas, set(concludableHas.asHas().owner(), concludableHas.asHas().attribute()));
+
+            return new ConjunctionAnswerCountEstimator.LocalEstimate.SimpleEstimate(list(concludableHas.asHas().owner(), concludableHas.asHas().attribute()), estimate);
+        }
+
+        private ConjunctionAnswerCountEstimator.LocalEstimate estimatesFromRolePlayersAssumingEvenDistribution(Concludable.Relation concludable) {
+            // Small inaccuracy: We double count duplicate roles (r:$a, r:$b)
+            // counts the case where r:$a=r:$b, which TypeDB wouldn't return
+            Set<RelationConstraint.RolePlayer> rolePlayers = concludable.relation().players();
+            List<Variable> constrainedVars = new ArrayList<>();
+            Map<TypeVariable, Long> rolePlayerEstimates = new HashMap<>();
+            Map<TypeVariable, Integer> rolePlayerCounts = new HashMap<>();
+
+            double relationTypeEstimate = countPersistedThingsMatchingType(concludable.relation().owner());
+
+            for (RelationConstraint.RolePlayer rp : rolePlayers) {
+                constrainedVars.add(rp.player());
+                TypeVariable key = rp.roleType().orElse(null);
+                rolePlayerCounts.put(key, rolePlayerCounts.getOrDefault(key, 0) + 1);
+                if (!rolePlayerEstimates.containsKey(key)) {
+                    rolePlayerEstimates.put(key, countPersistedRolePlayers(rp));
+                }
+            }
+
+            // TODO: Can improve estimate by collecting List<List<LocalEstimate>> from the triggered rules and doign sum(costCover).
+            // Consider owner in the inferred estimate call only if it's not anonymous
+            if (concludable.relation().owner().id().isName()) {
+                constrainedVars.add(concludable.relation().owner());
+            }
+            long inferredRelationsEstimate = estimateInferredAnswerCount(concludable, new HashSet<>(constrainedVars));
```

Comment:
Assumption made to keep things local. We can't have role-based projection without propagating constraints from the rules (possible, but likely overkill)